### PR TITLE
Feature/error handling

### LINKED ISSUE #24: Add error handling to components holding state
- [x] App
- [x] CommentsContainer
- [x] CommentForm
- [x] Error
- [x] TeaPage

### LINKED ISSUE #27: Add loading icon while fetching data

### PR DESCRIPTION
**Pull request by stephanieguzm**
# Description
Create Error component
Add error handling and conditional rendering to components: App, CommentsContainer, CommentForm, TeaPage
Add spinner loading icon for fetch delays

## Type of change

- [ ] Bug fix (fixes an issue)
- [x] New feature (adds functionality)
- [x] Refactor

# How Has This Been Tested?
Changes were tested in the browser and will be fully tested with Cypress in an upcoming iteration

- [x] I tested my changes in the browser
- [ ] I tested with Mocha/Chai
- [ ] I tested with Cypress

# Checklist:

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new bugs
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Is your code D.R.Y?
- [ ] Does it follow SRP?

Closes #24 
Closes #27

### CODE REVIEW

File: src/apiCalls.js [hunk 1]
```diff
@@ -28,7 +28,7 @@ const postComment = (newComment) => {
   })
   .then((resp) => {
     if (!resp.ok) {
-      throw new Error(resp.statusCode)
+      throw new Error(`${resp.statusCode}: Oops, we encountered an issue submitting your comment. Please try again.`)
```
Comment (yiranli624) [inline src/apiCalls.js]: all above error messages are looking like client side facing, it's fine as long as you understand users are not going to see these messages. They would only show up in devtools.

File: src/components/App/App.css [hunk 2]
```diff
@@ -20,4 +20,28 @@
...
 a:link {
   text-decoration: none;
+}
+
+.spinner {
+  margin: auto;
+  /* margin-left: 6em;
+  margin-right: 6em; */
+  display: block;
+  /* float: left; */
```
Comment (yiranli624) [inline src/components/App/App.css:30]: need clean up?

File: src/components/App/App.css [hunk 3]
```diff
@@ -20,4 +20,28 @@
...
+  border: 25px solid rgb(0, 0, 0);
+  width: 1px;
+  height: 1px;
+  border-radius: 50px;
+  border-left-color: transparent;
+  border-right-color: transparent;
+  animation-name: spin;
+  animation-duration: 1.5s;
+  animation-iteration-count: infinite;
+  animation-timing-function: ease-in-out;
```
Comment (yiranli624) [inline src/components/App/App.css]: nit: i think there's a shortcut to combine the animation

File: src/components/App/App.js [hunk 4]
```diff
@@ -13,35 +14,53 @@ class App extends Component {
...
-          />
-          <Route 
-            exact path='/' 
-            render={() => {
-              return <TeaContainer teas={this.state.teas}/>
-            }}
-          />
-        </Switch>
+        <div className='components-container'>
+          {!this.state.teas.length
```
Comment (yiranli624) [inline src/components/App/App.js]: nit: a positive condition might be easier to read

File: src/components/App/App.js [hunk 5]
```diff
@@ -13,35 +14,53 @@ class App extends Component {
...
-            }}
-          />
-        </Switch>
+        <div className='components-container'>
+          {!this.state.teas.length 
+          ? <Error 
+            errorMessage={this.state.error} 
+            returnHome={this.returnHome} />
+          : <Switch>
+            <Route
```
Comment (yiranli624) [inline src/components/App/App.js:38]: missing a tab before `<Route>`

File: src/components/App/App.js [hunk 6]
```diff
@@ -13,35 +14,53 @@ class App extends Component {
...
+            />
+            <Route 
+              exact path='/:id'
+              render={({ match }) => {
+                const teaId = match.params.id
+                return <TeaPage 
+                  teas={this.state.teas} 
+                  selectedTea={parseInt(teaId)} 
+                  returnHome={this.returnHome}
+                  errorMessage={this.state.error} />
```
Comment (yiranli624) [inline src/components/App/App.js]: this whole `match` part could be moved to `<TeaPage>` component. no need to figure out id at this level

File: src/components/CommentForm/CommentForm.js [hunk 7]
```diff
@@ -44,24 +49,29 @@ class CommentForm extends Component {
...
+          />
         <input 
           type='textarea'
           name='user_message'
           placeholder='Share your thoughts here'
           value={this.state.user_message}
           onChange={(event) => this.handleChange(event)}
-        />
-      
+          />
```
Comment (yiranli624) [inline src/components/CommentForm/CommentForm.js:67]: the original indentation looks right

File: src/apiCalls.js [hunk 8]
```diff
@@ -28,7 +28,7 @@ const postComment = (newComment) => {
...
     }
     console.log(resp)
```
Comment (yiranli624) [inline src/apiCalls.js]: leftover console?

File: src/components/TeaPage/TeaPage.js [hunk 9]
```diff
@@ -1,48 +1,63 @@
...
+  //       }
+  //       return response.json() 
+  //     })
+  //     .then( data => this.setState({ tea: data[0] }))
+  //     .catch( error => {
+  //       if (error) {
+  //         this.setState({ error: true })
+  //       }
+  //     })
+  // }
```
Comment (yiranli624) [inline src/components/TeaPage/TeaPage.js:34]: leftover cleanup?

File: src/components/TeaPage/TeaPage.js [hunk 10]
```diff
@@ -1,48 +1,63 @@
...
-      .catch( error => {
-        if (error) {
-          this.setState({ error: true })
-        }
-      })
+
+  componentDidMount = () => {
+    const matchingTea = this.props.teas.find( tea => tea.id === this.props.selectedTea)
+    this.setState({tea: matchingTea})
+    console.log(matchingTea)
```
Comment (yiranli624) [inline src/components/TeaPage/TeaPage.js]: leftover console

File: src/components/CommentsContainer/CommentsContainer.js [hunk 11]
```diff
@@ -28,15 +29,24 @@ class CommentsContainer extends Component {
 
   render() {
     return (
-      <div className='comments-container'>
+      <>
```
Comment (yiranli624) [inline src/components/CommentsContainer/CommentsContainer.js:32]: this empty div doesnt seem necessary. The following section has wrong indentation

File: src/components/TeaPage/TeaPage.js [hunk 12]
```diff
@@ -1,48 +1,63 @@
...
-          <h3>{tea.name}</h3>
-          <p>{tea.description}</p>
-          <p>Originated from {tea.origin}</p>
-          <h4>Make the perfect cup</h4>
-          <p>Brew at {tea.temperature} for {tea.brewTime} minutes</p>
-          <img src={tea.img} alt={tea.name}/>
-        </div>
-        <div className='comments-container'>
-          <CommentsContainer tea_id={this.props.selectedTea}/>
+      <>
```
Comment (yiranli624) [inline src/components/TeaPage/TeaPage.js:39]: same here. not useful empty div and incorrect indentation following up

File: src/apiCalls.js (hunk 1 above)
Comment (stephanieguzm) [reply src/apiCalls.js]: You're right! I will make those changes on the calls.

File: src/components/App/App.css (hunk 2 above)
Comment (stephanieguzm) [reply src/components/App/App.css:30]: I'm still toying with some styling here so these will stay.

File: src/components/App/App.css (hunk 3 above)
Comment (stephanieguzm) [reply src/components/App/App.css]: You're right! This will be combined under animation.

File: src/components/App/App.js (hunk 5 above)
Comment (stephanieguzm) [reply src/components/App/App.js:38]: Spacing is correct, just looks off because of colon.

File: src/components/App/App.js (hunk 4 above)
Comment (stephanieguzm) [reply src/components/App/App.js]: Changed condition to positive

File: src/apiCalls.js (hunk 8 above)
Comment (stephanieguzm) [reply src/apiCalls.js]: Removed

File: src/components/CommentsContainer/CommentsContainer.js (hunk 11 above)
Comment (stephanieguzm) [reply src/components/CommentsContainer/CommentsContainer.js:32]: Good point but in this case I need the fragment to hold my conditional rendering. Fixed indentation.

File: src/components/TeaPage/TeaPage.js (hunk 9 above)
Comment (stephanieguzm) [reply src/components/TeaPage/TeaPage.js:34]: Yes, I am still needing to convert this component into a functional one so this will stay for now.

**Pull request merged**